### ISSUE TITLE
definition of paths over may need a universe

### ISSUE THREAD

**Issue by DanGrayson** [UniMath/SymmetryBook#121]
Here is my definition of paths over a path:
<img width="574" alt="Screen Shot 2021-08-20 at 12 01 28 PM" src="https://user-images.githubusercontent.com/700228/130261436-268f67d0-3dbd-47a7-9dd9-30c735814e12.png">

And here is path induction:
<img width="589" alt="Screen Shot 2021-08-20 at 12 03 19 PM" src="https://user-images.githubusercontent.com/700228/130261539-6b2998ec-57d4-4d6e-8e4b-3c296a1c23fa.png">

So we need a suitable universe to contain everything.  But this is long before the section on universes.  And in the section on universes we don't even say that every family `Y(x)` of types has all its members contained in some universe.

Or we need to revert to the asymmetric definition involving transport.

Or we need a new form of path induction devoted to defining new types.  This seems to be the cleanest way to proceed.

**Comment by UlrikBuchholtz** [comment]
Yes, before we have universes we need all inductive types, including path types, to support large elimination.

The question is how best to deal with this pedagogically. Maybe the best option is just to face it head on, and mention it as an additional rule for each inductive type. For natural numbers, for example, we can mention the family of higher type functionals, F 0 = N, F (Succ n) = F n → N.

But then, after we introduce universes, large elimination comes for free, so this should be noted.

> And in the section on universes we don't even say that every family `Y(x)` of types has all its members contained in some universe.

We do, at the beginning of Sec. 2.12 on universes, where we say that we “make precise” the notion of a family of types by taking it to be a function to a universe.

**Comment by DanGrayson** [comment]
The previous notion of a family of types was already precise, and what this section does is to *change* the definition of the notion:

<img width="576" alt="Screen Shot 2021-08-20 at 4 23 49 PM" src="https://user-images.githubusercontent.com/700228/130289742-fe699b61-62b8-4bec-bbbc-5f3785f6891a.png">

I like the idea of adding additional inductive rules -- they're easy to motivate.

**Comment by DanGrayson** [comment]
Re:

> But then, after we introduce universes, large elimination comes for free, so this should be noted.

Not totally for free -- one must still prove that each family of types has every member in some universe.  And that's a meta-mathematical statement, so we should avoid delving into it.

**Comment by DanGrayson** [comment]
Hmm, would we need a totally new identity type whose elements identify 2 types?

(So we could use it here:

<img width="577" alt="Screen Shot 2021-08-20 at 4 32 17 PM" src="https://user-images.githubusercontent.com/700228/130290562-b845cf5f-a258-4c61-b2e0-b8da3a51c5d9.png">
)

**Comment by UlrikBuchholtz** [comment]
> Hmm, would we need a totally new identity type whose elements identify 2 types?

Yes, we would. And, on p. 25 (in Sec. 2.12), we would then need to add: (6') If X,X' : U, then (X = X') : U and (X = X') ≡ (X =_U X') : U', when U : U'. Or only: (X = X') : U', as the former option makes the identity types in U U-small on the nose, and not only essentially U-small, which is what we get from univalence.

And we'd probably need definitional equalities relating large eliminations to U-small eliminations as well. (Maybe in App. B we could/should say something about universes à la Tarski/Russell, cumulativity, etc.)

My worry is that these issues will be a distraction in Ch. 2. There are also some issues that may still be subject to ongoing research. For example, I don't know/remember if there's been progress on conservativity of Russell-style over Tarksi-style since Zhaohui Luo's [Notes on Universes](http://www.cs.rhul.ac.uk/home/zhaohui/universes.pdf).

**Comment by DanGrayson** [comment]
I don't see how to provide a semantic interpretation for X=X' in topology -- you need a universe U to speak of U-small fibrations over [0,1] connecting X to X'.  I think we'd better stay away from all that.  We could move the section on universes so it comes before this section.  Or we could instead take one of the two asymmetric definitions for paths over involving transport.

**Comment by UlrikBuchholtz** [comment]
Yes, maybe it's better pedagogically to just switch to the transport definition.

BTW, in Sec. 2.12 on universes, shouldn't we rephrase item (5) on p. 25 to include all inductive types? (And any higher inductive types we may later introduce.)

(Re semantics, I don't see the problem: If nothing else, we could define X=X' to *be* the type of equivalences X≃X', which itself is a subtype of X→X'.)

**Comment by DanGrayson** [comment]
re 2.12: we haven't introduced a general scheme for making new inductive definitions, so what do we mean by "all inductive types".

re semantics: take that approach and you probably can't validate the existence of a map between X=X' and X =_U X', in either direction.

**Comment by DanGrayson** [comment]
I'll switch to the transport definition.

**Comment by UlrikBuchholtz** [comment]
Great!

> re semantics: take that approach and you probably can't validate the existence of a map between X=X' and X =_U X', in either direction.

Isn't that exactly univalence?

**Comment by DanGrayson** [comment]
Oh, right -- there should be a map X =_U X' ---> X = X', and it should be an equivalence.  Now I'm suspicious of transport by elements of X = X'.

**Comment by UlrikBuchholtz** [comment]
> re 2.12: we haven't introduced a general scheme for making new inductive definitions, so what do we mean by "all inductive types"?

At the very least, we could say “all the particular ones introduced hitherto (e.g., in Sec. 2.11), as well as any others that will be introduced later on.”

BTW, I just noticed: We use large eliminations twice on p. 23: Once for the family P over Bool and again in Ex. 2.11.3 for the family T (again over Bool).

**Comment by UlrikBuchholtz** [comment]
We say now, on p. 23, “Thirdly, there will be a type called Bool, defined by induction and provided with two elements, yes and no.”

Bool isn't defined by induction – rather, it's an inductive type!

**Comment by DanGrayson** [comment]
Instead of introducing that terminology, it might be good just to display all the rules for each of the types in that section.

**Comment by UlrikBuchholtz** [comment]
> Instead of introducing that terminology, it might be good just to display all the rules for each of the types in that section.

You mean “large elimination”? Yes, that terminology can be relegated to the margin. I don't think we need to give the exact rule for each inductive type; it should suffice to spell it out in Sec. 2.3 (on natural numbers), make a note about it in Sec. 2.4 (on identity types), and then another note in Sec. 2.11, perhaps in the introductory paragraph. But what did you have in mind?

**Comment by DanGrayson** [comment]
No, I was thinking of "inductive type", which might be hard to define.  Well, maybe not -- we could just say that inductive types are the ones we present, which are characterized by the rules for using them, one of which is for producing elements by induction.

As for "large elimination", we probably should not mention it, nor use it.

I was thinking that for empty, unit, and bool, we should spell out, for each of them, all the rules.  Otherwise we're asking students to guess.

**Comment by UlrikBuchholtz** [comment]
Well we use the term “inductive type” many times, even in the title of Sec. 2.11.

If we want to do without large eliminations, then we do need to move the universe section to prior to Sec. 2.11, or move the bits about proving yes ≠ no and 0 ≠ 1 etc. to the universe section.

**Comment by DanGrayson** [comment]
On second thought, it might be sort of cool to have large elimination rules for each of our inductive types.  Well, maybe not for the circle.

**Comment by UlrikBuchholtz** [comment]
Why not? A key point in Ch. 3 is to consider type families over the circle as types with a self-equivalence.

**Comment by DanGrayson** [comment]
The question for me is whether anyone is including large elimination for the circle in a sound semantic model based on topology.

**Comment by UlrikBuchholtz** [comment]
I'm no expert on semantics, but don't we usually work in a meta-theory where any type/set belongs to some (Grothendieck) universe. If A → S¹ is a fibration, then A (and S¹) belong to some universe U, and the fibration itself then lives in the semantics of U-small type families over S¹.

If there is anything to worry about, it's probably cumulativity of universes: As I mentioned above, that's more likely to cause trouble either semantically and/or for canonicity.

**Comment by DanGrayson** [comment]
If you model X =_U X' with U-small fibrations over [0,1], then the thing depends on U, so you can't expect to get a well defined object for X = X' by choosing a universe at random.

**Comment by UlrikBuchholtz** [comment]
I think that illustrates rather why one should be dubious about cumulativity in general.

I was suggesting above to model X = X' as X ≃ X' which is patently universe-invariant. The question then re-appeared in the guise of large elimination for HITs. I just had a look at [Lumsdaine–Shulman](https://arxiv.org/abs/1705.07088), and they *only* do large eliminations, so that's in a sense easier than ensuring the universes are closed under HITs.

**Comment by UlrikBuchholtz** [comment]
Sorry, that's not right: They say you need universes to get large eliminations. But I still think they're easier to get than cumulative universes, though as I said, I'm no expert on semantics.

**Comment by DanGrayson** [comment]
I was worried above about modeling transport along an identity of type X = X', but transport is only for identities between two elements of a type.  Let's just stay away from identities between two types, unless we are already thinking of them as elements of a single universe.  I can rephrase 2.6.2 to produce an equivalence between the two possible definitions of paths-over.

As for cumulative universes, Vladimir preferred the Tarski style.

**Comment by marcbezem** [comment]
There are 57 lines on with the macro \pathover occurs, 36 of which are in intro-uf.tex.
There may also be some pictures involved.
As an alternative, we could move the definition till after universes are introduced.
Let's discuss this on one of our Thursday meetings.

**Comment by bidundas** [comment]
As you know, I’m totally fine with scrapping pathovers.

Where, if anywhere, have we written X=X’ when X and X’ are not in some (probably unspecified) universe U with a different interpretation than X=_UX’ ?


Bjorn

**Comment by DanGrayson** [comment]
I used X = X' here:

<img width="577" alt="130290562-b845cf5f-a258-4c61-b2e0-b8da3a51c5d9" src="https://user-images.githubusercontent.com/700228/130451463-db987ab5-b905-4ec6-bc48-19303047e4b9.png">

**Comment by DanGrayson** [comment]
... but it's not essential -- the identity can be converted to an equivalence.

**Comment by DanGrayson** [comment]
> As you know, I’m totally fine with scrapping pathovers. 

I really like the way they make this result look, though:

<img width="320" alt="Screen Shot 2021-08-23 at 9 49 12 AM" src="https://user-images.githubusercontent.com/700228/130458747-855bf619-7921-4893-b2f7-fecc0e1cfe98.png">

**Comment by DanGrayson** [comment]
> How about the following:
> 
> ```
> (E4) Suppose we are given: a type X and an element a : X;
> a family of types P(b,e) as in (E3);
> as well as a type family Tₐ(p) parametrized by p : P(a,refl).
> Then we get a family of types T(b,e,p) parametrized by b : X, e : a = b, and p : P(b,e).
> Moreover, T(a,refl,p) ≡ Tₐ(p) for all p:P(a,refl)
> ```
> 
> In Def. 2.6.1 we then take P(b,e) :≡ Y(b) and Tₐ(y') :≡ (y = y) for y' : Y(a).

That covers just one special case, so it doesn't seem general enough.  Somehow (E3) "works" without even mentioning the possibility of further items in the context.  Of course, some students might notice that problem with (E3).

**Comment by DanGrayson** [comment]
> But it doesn't matter, since we decided to do “Early universes”/“In the beginning was the universe hierarchy, and the universe hierarchy was with the types, and the types were classified by universes.”

Oh, okay.  I guess we'd better put it after 2.2 (Types, etc.) and 2.3 (natural numbers).  Then we'll be prepared to say that induction on a natural number.

I'd be happy to take care of it.

**Comment by bidundas** [comment]
“In the beginning was the universe hierarchy, and the universe hierarchy was with the types, and the types were classified by universes.”

I think the gospel according to Marc/Daniel/Ulrik/Pierre/Bjorn will be looked on kindly by Pope Francis, but I’m concerned about having a horde of evangelical fundamentalists at our door if we publish it in this form.

Bjorn

**Comment by DanGrayson** [comment]
Okay, I've made the relevant changes (except for the ones related to making type families over bool).  But now I'm sad:

<img width="513" alt="Screen Shot 2021-09-04 at 10 58 07 AM" src="https://user-images.githubusercontent.com/700228/132099192-a4e2d602-7f4e-436e-8727-3c4805aa7a07.png">

Do we have to address what happens to the path-over type if a different universe is chosen?

**Comment by UlrikBuchholtz** [comment]
That seems like the same sort of issue that causes problems with canonicity when we have cumulativity. I don't know how that's taken care of in Thierry's [paper](https://arxiv.org/abs/1810.09367), but we could ask him.

BTW, I changed the word “artifice” to “instrument” in the beginning of (the new) Sec. 2.3 to describe universes, as I don't want to give the impression that universes are inherently deceptive (although many surrounding issues certainly can be tricky!).

**Comment by UlrikBuchholtz** [comment]
What do you think of Thierry's suggestion to take dependent paths as a primitive? We can of course prove that it is equivalent to both the indexed inductive family and the two transport equalities.

**Comment by bidundas** [comment]
I was fine with sticking with transport (and as to the argument about asymmetry, the fact that the induction singles out one of the endpoints makes all definitions asymmetric), so I take this as a theological discussion from which I am exempted. 

Bjorn

**Comment by bidundas** [comment]
PS. I can’t come Thursday.

Bjorn

**Comment by marcbezem** [comment]
Dependent paths as a primitive in 2.3 is quite difficult for the reader. Isn't it better to pay some attention to the asymmetry later in the book (heading: "Asymmetry")? Ulrik mentions the equivalence of the two transport equalities. Bjørn is right that based path induction is already asymmetric, we could show that choosing "the other" base point is equivalent (or have an exercise to that effect).

**Comment by DanGrayson** [comment]
I've taken a stab at defining families of types by induction in commit 8f6716231cb48d867fdb5020855df48ae6fcb4b2.

Here is the new formulation of E3 and the new rule E4:

<img width="619" alt="Screen Shot 2021-09-23 at 9 02 48 AM" src="https://user-images.githubusercontent.com/700228/134521568-95305ede-23ce-40be-af53-07252bd01b0f.png">

And here is how paths over paths are now defined.

<img width="602" alt="Screen Shot 2021-09-23 at 9 02 07 AM" src="https://user-images.githubusercontent.com/700228/134521640-558958bf-6a98-4c39-9462-912ea7e02d3c.png">

This single commit is easy to revert, if we don't like this.

**Comment by DanGrayson** [comment]
Here's a new paragraph explaining that a family may depend on more than one variable:

<img width="617" alt="Screen Shot 2021-09-23 at 9 23 32 AM" src="https://user-images.githubusercontent.com/700228/134525252-a4d8abe4-c217-4684-b756-f6bb2b0f3617.png">

**Comment by DanGrayson** [comment]
Here's a slightly improved version of E4 as suggested by Ulrik:

<img width="617" alt="Screen Shot 2021-09-23 at 9 27 01 AM" src="https://user-images.githubusercontent.com/700228/134525879-aedbec13-60af-4060-9cfc-e9baf07a1dff.png">

**Comment by DanGrayson** [comment]
A serious drawback to introducing an induction rule for a family of types is that Coq doesn't offer it, nor do other proof assistants, so we can't formalize the book exactly.

I suspect also that we don't get independence of the choice of universe definitionally in coq, but will check...

**Comment by DanGrayson** [comment]
I tried to check independence of the choice of universe definitionally in coq, but I'm defeated by the tricky handling of universe in coq -- I can't seem to get two instances of PathOver in the same statement to refer to different universes.

**Comment by DanGrayson** [comment]
> A serious drawback to introducing an induction rule for a family of types is that Coq doesn't offer it, nor do other proof assistants, so we can't formalize the book exactly.
>

On the other hand, Coq's handling of universes is all done behind the scenes, so it looks to the reader like Coq does offer it.  Here's the definition in UniMath, currently:

```coq
Definition PathOver {X:Type}
           {x x':X} (p:x=x')
           {Y : X -> Type} (y : Y x) (y' : Y x') : Type.
Proof.
  induction p. exact (y=y').
Defined.
```

**Comment by DanGrayson** [comment]
Our current definition is this:
<img width="566" alt="Screen Shot 2021-09-24 at 8 46 21 AM" src="https://user-images.githubusercontent.com/700228/134685028-ad65d818-9525-4ee2-9d22-7f639bf5eaf0.png">

And UniMath's current definition is this:
```coq
Definition PathOver {X:Type}
           {x x':X} (p:x=x')
           {Y : X -> Type} (y : Y x) (y' : Y x') : Type.
Proof.
  induction p. exact (y=y').
Defined.
```

Isn't that pleasant?

**Comment by UlrikBuchholtz** [comment]
> And UniMath's current definition is this:
> ```coq
> Definition PathOver {X:Type}
>            {x x':X} (p:x=x')
>            {Y : X -> Type} (y : Y x) (y' : Y x') : Type.
> Proof.
>   induction p. exact (y=y').
> Defined.
> ```
> Isn't that pleasant?

If you `Print PathOver` in Coq, you see that Coq inferred the use of J (`paths_rect`) with the motive Π(x':X)(p:x=x'), Y x' → Type, i.e., the type family over (x',p) is Y x' → Type. So it's a standard “small” elimination applied to a large motive (involving Type).

So I agree it's pleasant, and I'm just adding that it's captured by having universes (as in 2.3) and using the usual J eliminator (E3).

BTW, you asked what the rules would be for a “primitive” dependent paths type. I think they would be precisely the rules you'd get if you define it as an indexed inductive family:

* Formation: If X : Type, Y : X → Type, x,x' : X, p : x=x', y : Y x, y' : Y x', then y =^Y_p y' : Type
* Introduction rule: We have reflo : y =^Y_refl y for all x,y.
* Elimination rule: Fixing X,Y,x,y, if we have M : Π(x',p,y'), y =^Y_p y' → Type and m : M(x,refl,y,reflo), then we get a section J' of M.
* Computation rule: J'(x,refl,y,reflo) = m definitionally.

The downside of this (and presumably all other “primitive” versions, not counting anything in cubical type theory) is that we don't get (y =^Y_refl y') = (y = y') definitionally, but only via an equivalence. This is the benefit of both the inductive definition we have now, and the definition (y =^Y_p y') := (transport Y p y = y').

**Comment by DanGrayson** [comment]
Yes, it is captured by the usual J eliminator, but Coq's strategy of using symbolic universe levels when you refer to "Type", and solving for the levels at the end of the game, seems to convey the idea that it really is large elimination, since the universe can grow when necessary.

**Comment by DanGrayson** [comment]
We already have families of types, since they occur in making Pi-types, so it would be really nice to have more ways to make them.  And if the models really support universe-independent "large" "elimination", it adds elegance to the formulation to take advantage of it.
